broker/ga4gh/broker/endpoints/builds.py

```diff
@@ -0,0 +1,604 @@
+    """Register new builds for already registered repository.
+
+    Args:
+        repository_id: Identifier for repository.
+        access_token: Secret used to verify source of the request to
+        initiate new build.
+        build_data: Request data containing build information.
+
+    Returns:
+        build_id: Identifier for new registered build.
+
+    Raises:
+        Unauthorized: Raised when access_token is invalid or not specified
+        in request.
+        RepositoryNotFound: Raised when repository is not found with given
+        identifier.
+
+    Description:
+        - Takes request build_data (Build information).
+        - Check if repository with specified identifier is available or not.
+        - Verify access_token.
+        - Generate a random string, add it after repository identifier and
+        use it as build identifier. (Retries 3 times for generating unique
+        random string.)
+        - Insert build_data in mongodb.
+        - Call create_build function.
+        - returns build_id.
+    """
+    retries = 3
+    base_dir = '/broker_temp_files'
+    db_collection_builds = (
+        current_app.config['FOCA'].db.dbs['brokerStore'].
+        collections['builds'].client
+    )
+    db_collection_repositories = (
+        current_app.config['FOCA'].db.dbs['brokerStore'].
+        collections['repositories'].client
+    )
+    id_length = (
+        current_app.config['FOCA'].endpoints['repository']['id_length']
+    )
+    id_charset: str = (
+        current_app.config['FOCA'].endpoints['repository']['id_charset']
+    )
+    try:
+        id_charset = eval(id_charset)
+    except Exception:
+        id_charset = ''.join(sorted(set(id_charset)))
+
+    data_from_db = db_collection_repositories.find_one({'id': repository_id})
+    if data_from_db is None:
+        logger.error(
+                    f"Could not find repository with given identifier: " +
+                    repository_id
+                )
+        raise RepositoryNotFound
+    if data_from_db['access_token'] != access_token:
+        raise Unauthorized
+    for i in range(retries + 1):
+        logger.debug(
+            f"Trying to insert/update object: try {i}" +
+            str(build_data))
+        build_data['id'] = repository_id + generate_id(
+            charset=id_charset,
+            length=id_length,
+        )
+        db_collection_repositories.update_one({"id": repository_id},
+                                              {"$push": {
+                                               "build_list":
+                                                   build_data['id']}})
+        try:
+            build_data['finished_at'] = "NULL"
+            build_data['started_at'] = str(
+                datetime.datetime.now().isoformat())
+            build_data['status'] = "QUEUED"
+            db_collection_builds.insert_one(build_data)
+            if 'branch' in build_data['head_commit'] and 'commit_sha' in \
```

Comment:
Wouldn't it be easy to first get the value of commit in a variable, and then call `create_build` instead of calling the function 3 times?

Avery python way to achieve what you try to do is:

```
try:
    build_data['head_commit']['branch']
except KeyError:
    build_data=''
```

Also there is no `else` clause, are you sure there is no way to have something different that `tag` or `branch`? I guess Swagger takes care of this check.